File: concourse-driver-java/src/main/java/org/cinchapi/concourse/Link.java
/*
 * Copyright (c) 2013-2015 Cinchapi, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.cinchapi.concourse;

import javax.annotation.concurrent.Immutable;

import com.google.common.primitives.Longs;
import com.google.common.primitives.UnsignedLongs;

/**
 * A {@link Link} is a wrapper around a {@link Long} that represents the primary
 * key of a record and distinguishes from simple long values. A Link is
 * returned from the {@code #fetch} or {@code #get} methods in {@link Concourse}
 * if data was added using one of the {@code #link} operations.
 * 
 * @author Jeff Nelson
 */
// NOTE: This class extends Number so that it can be treated like other
// numerical Values during comparisons when the data is stored and sorted in a
// SecondaryIndex for efficient range queries.
@Immutable
public final class Link extends Number implements Comparable<Link> {

    /**
     * Return a Link that points to {@code record}.
     * 
     * @param record
     * @return the Link
     */
    public static Link to(long record) {
        return new Link(record);
    }

    private static final long serialVersionUID = 1L; // Serializability is
                                                     // inherited from {@link
                                                     // Number}.
    /**
     * The signed representation for the primary key of the record to which this
     * Link points.
     */
    private final long record;

    /**
     * Construct a new instance.
     * 
     * @param record
     */
    private Link(long record) {
        this.record = record;
    }

    @Override
    public int compareTo(Link other) {
        return UnsignedLongs.compare(longValue(), other.longValue());
    }

    @Override
    public double doubleValue() {
        return (double) record;
    }

    @Override
    public boolean equals(Object obj) {
        if(obj instanceof Link) {
            Link other = (Link) obj;
            return UnsignedLongs.compare(record, other.record) == 0;
        }
        return false;
    }

    @Override
    public float floatValue() {
        return (float) record;
    }

    @Override
    public int hashCode() {
        return Longs.hashCode(record);
    }

    @Override
    public int intValue() {
        return (int) record;
    }

    @Override
    public long longValue() {
        return record;
    }

    @Override
    public String toString() {
        return UnsignedLongs.toString(longValue()); // for
                                                    // compatibility
                                                    // with
                                                    // {@link
                                                    // com.cinchapi.common.Numbers.compare(Number,
                                                    // Number)}
    }
}

File: concourse-driver-ruby/lib/concourse/link.rb
# Copyright (c) 2013-2017 Cinchapi Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module Concourse

    # A Link is a pointer to a record.
    #
    # Links should never be written directly. They can be created using the
    # `Concourse.link` method in the `Concourse` API.
    #
    # Links may be returned when reading data using the `Concourse.select()`,
    # `Concourse.get()` and `Concourse.browse()` methods. When handling Link
    # objects, you can retrieve the underlying record id by accessing the
    # `Link.record` property.
    #
    # When performing a bulk insert (using the `Concourse.insert()` method) you
    # can use this class to create Link objects that are added to the data/json
    # blob. Links inserted in this manner will be written in the same way they
    # would have been if they were written using the `Concourse.link()` method.
    #
    # To create a static link to a single record, use `Link.to()`.
    #
    # To create static link to each of the records that match a criteria, use
    # the `Link.to_where` method.
    class Link

        # Allow read access to the _record_ field.
        attr_reader :record

        # Alias for the constructor.
        # @param [Integer] record The target of the link
        # @return [Link] the Link
        def self.to record
            Link.new record
        end

        # Return a string that instructs Concourse to create links that point to
        # each of the records that match the _ccl_ string.
        #
        # NOTE: This method does not return a _Link_ object, so it should only
        # be used when adding a resolvable link value to a data/json blob that
        # will be passed to the `Concourse.insert()` method.
        #
        # @param [String] ccl a CCL string that describes the records to which a
        #   Link should point
        # @return [String] a resolvable link instruction
        def self.to_where ccl
            "@#{ccl}@"
        end

        # Return a _Link_ that points to _record_.
        #
        # @param [Integer] record The record id
        # @return [Link] a _Link_ that points to _record_
        def initialize record
            @record = record
        end

        # Overriden
        def to_s
            "@#{@record}"
        end

        # Overriden
        def ==(other)
            if other.is_a? Link
                return other.record == record
            else
                return false
            end
        end

    end
end
